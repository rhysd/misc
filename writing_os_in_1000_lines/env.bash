#!/bin/bash

export PATH="/usr/local/opt/llvm/bin:$PATH"
