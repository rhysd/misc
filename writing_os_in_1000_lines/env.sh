#!/bin/sh

export PATH="/usr/local/opt/llvm/bin:$PATH"
