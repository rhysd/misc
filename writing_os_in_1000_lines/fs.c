#include "kernel.h"

#define SECTOR_SIZE 512
#define VIRTQ_ENTRY_NUM 16
#define VIRTIO_DEVICE_BLK 2
#define VIRTIO_REG_MAGIC 0x00
#define VIRTIO_REG_VERSION 0x04
#define VIRTIO_REG_DEVICE_ID 0x08
#define VIRTIO_REG_QUEUE_SEL 0x30
#define VIRTIO_REG_QUEUE_NUM_MAX 0x34
#define VIRTIO_REG_QUEUE_NUM 0x38
#define VIRTIO_REG_QUEUE_ALIGN 0x3c
#define VIRTIO_REG_QUEUE_PFN 0x40
#define VIRTIO_REG_QUEUE_READY 0x44
#define VIRTIO_REG_QUEUE_NOTIFY 0x50
#define VIRTIO_REG_DEVICE_STATUS 0x70
#define VIRTIO_REG_DEVICE_CONFIG 0x100
#define VIRTIO_STATUS_ACK 1
#define VIRTIO_STATUS_DRIVER 2
#define VIRTIO_STATUS_DRIVER_OK 4
#define VIRTIO_STATUS_FEAT_OK 8
#define VIRTQ_DESC_F_NEXT 1
#define VIRTQ_DESC_F_WRITE 2
#define VIRTQ_AVAIL_F_NO_INTERRUPT 1
#define VIRTIO_BLK_T_IN 0
#define VIRTIO_BLK_T_OUT 1
#define FILES_MAX 3
#define DISK_MAX_SIZE align_up(sizeof(struct file) * FILES_MAX, SECTOR_SIZE)

struct virtq_desc {
    uint64_t addr;
    uint32_t len;
    uint16_t flags;
    uint16_t next;
} __attribute__((packed));

struct virtq_avail {
    uint16_t flags;
    uint16_t index;
    uint16_t ring[VIRTQ_ENTRY_NUM];
} __attribute__((packed));

struct virtq_used_elem {
    uint32_t id;
    uint32_t len;
} __attribute__((packed));

struct virtq_used {
    uint16_t flags;
    uint16_t index;
    struct virtq_used_elem ring[VIRTQ_ENTRY_NUM];
} __attribute__((packed));

struct virtio_virtq {
    struct virtq_desc descs[VIRTQ_ENTRY_NUM];
    struct virtq_avail avail;
    struct virtq_used used __attribute__((aligned(PAGE_SIZE)));
    int queue_index;
    volatile uint16_t *used_index;
    uint16_t last_used_index;
} __attribute__((packed));

// 5.2.6.4 Legacy Interface: Framing Requirements
// > MUST use a single 8-byte descriptor containing type, reserved and sector, followed by descriptors
// > for data, then finally a separate 1-byte descriptor for status.
struct virtio_blk_req {
    // The first descriptior. Read-only from virtio-blk device.
    uint32_t type;
    uint32_t reserved;
    uint64_t sector;
    // The second descriptor (VIRTQ_DESC_F_WRITE).
    uint8_t data[512];
    // The third descriptor (VIRTQ_DESC_F_WRITE)
    uint8_t status;
} __attribute__((packed));

// UStar header: https://en.wikipedia.org/wiki/Tar_(computing)#UStar_format
struct tar_header {
    char name[100];
    char mode[8];
    char uid[8];
    char gid[8];
    char size[12];
    char mtime[12];
    char checksum[8];
    char type;
    char linkname[100];
    char magic[6];
    char version[2];
    char uname[32];
    char gname[32];
    char devmajor[8];
    char devminor[8];
    char prefix[155];
    char padding[12];
    char data[]; // Pointer to the data after this header
} __attribute__((packed));

static struct virtio_virtq *blk_request_vq;
static struct virtio_blk_req *blk_req;
static paddr_t blk_req_paddr;
static unsigned blk_capacity;
static struct file files[FILES_MAX];
static uint8_t disk[DISK_MAX_SIZE];

static uint32_t virtio_reg_read32(unsigned const offset) {
    return *((volatile uint32_t *)(VIRTIO_BLK_PADDR + offset));
}

static uint64_t virtio_reg_read64(unsigned const offset) {
    return *((volatile uint64_t *)(VIRTIO_BLK_PADDR + offset));
}

static void virtio_reg_write32(unsigned const offset, uint32_t const value) {
    *((volatile uint32_t *)(VIRTIO_BLK_PADDR + offset)) = value;
}

static void virtio_reg_fetch_and_or32(unsigned const offset, uint32_t const value) {
    virtio_reg_write32(offset, virtio_reg_read32(offset) | value);
}

// Initialize virtual queue
struct virtio_virtq *virtq_init(unsigned const index) {
    // 4.2.4 Legacy interface
    paddr_t virtq_paddr = alloc_pages(align_up(sizeof(struct virtio_virtq), PAGE_SIZE) / PAGE_SIZE);
    struct virtio_virtq *vq = (struct virtio_virtq *)virtq_paddr;
    vq->queue_index = index;
    vq->used_index = (volatile uint16_t *)&vq->used.index;
    // 1. Select the queue writing its index (first queue is 0) to QueueSel.
    virtio_reg_write32(VIRTIO_REG_QUEUE_SEL, index);
    // 5. Notify the device about the queue size by writing the size to QueueNum.
    virtio_reg_write32(VIRTIO_REG_QUEUE_NUM, VIRTQ_ENTRY_NUM);
    // 6. Notify the device about the used alignment by writing its value in bytes to QueueAlign.
    virtio_reg_write32(VIRTIO_REG_QUEUE_ALIGN, 0);
    // 7. Write the physical number of the first page of the queue to the QueuePFN register.
    virtio_reg_write32(VIRTIO_REG_QUEUE_PFN, virtq_paddr);
    return vq;
}

// Explanation of Virtio buffer operation: https://www.redhat.com/ja/blog/virtqueues-and-virtio-ring-how-data-travels

// Initialize virtio-blk device
void virtio_blk_init(void) {
    if (virtio_reg_read32(VIRTIO_REG_MAGIC) != 0x74726976) {
        PANIC("virtio: invalid magic value");
    }
    if (virtio_reg_read32(VIRTIO_REG_VERSION) != 1) {
        PANIC("virtio: invalid version");
    }
    if (virtio_reg_read32(VIRTIO_REG_DEVICE_ID) != VIRTIO_DEVICE_BLK) {
        PANIC("virtio: invalid device id");
    }

    // 3.1.1 Driver Requirements: Device Initialization The driver MUST follow this sequence to initialize a device:

    // 1. Reset the device.
    virtio_reg_write32(VIRTIO_REG_DEVICE_STATUS, 0);
    // 2. Set the ACKNOWLEDGE status bit: the guest OS has noticed the device.
    virtio_reg_fetch_and_or32(VIRTIO_REG_DEVICE_STATUS, VIRTIO_STATUS_ACK);
    // 3. Set the DRIVER status bit.
    virtio_reg_fetch_and_or32(VIRTIO_REG_DEVICE_STATUS, VIRTIO_STATUS_DRIVER);
    // 5. Set the FEATURES_OK status bit.
    virtio_reg_fetch_and_or32(VIRTIO_REG_DEVICE_STATUS, VIRTIO_STATUS_FEAT_OK);
    // 7. Perform device-specific setup, including discovery of virtqueues for the device
    blk_request_vq = virtq_init(0);
    // 8. Set the DRIVER_OK status bit.
    virtio_reg_write32(VIRTIO_REG_DEVICE_STATUS, VIRTIO_STATUS_DRIVER_OK);

    // Get the disk space
    blk_capacity = virtio_reg_read64(VIRTIO_REG_DEVICE_CONFIG + 0) * SECTOR_SIZE;
    printf("virtio-blk: capacity is %d bytes\n", blk_capacity);

    // Allocate memory for processing device request
    blk_req_paddr = alloc_pages(align_up(sizeof(*blk_req), PAGE_SIZE) / PAGE_SIZE);
    blk_req = (struct virtio_blk_req *)blk_req_paddr;
}

// Notify virtio-blk device that there is a new request. desc_index is an index of the first descriptor
// of the new request.
static void virtq_kick(struct virtio_virtq *vq, int const desc_index) {
    vq->avail.ring[vq->avail.index % VIRTQ_ENTRY_NUM] = desc_index;
    vq->avail.index++;
    __sync_synchronize();
    virtio_reg_write32(VIRTIO_REG_QUEUE_NOTIFY, vq->queue_index);
    vq->last_used_index++;
}

// Check if there is some request being processed by the virtio-blk device.
static bool virtq_is_busy(struct virtio_virtq const *vq) {
    return vq->last_used_index != *vq->used_index;
}

// Read/Write virtio-blk device
static void read_write_disk(void *buf, unsigned const sector, int const is_write) {
    if (sector >= blk_capacity / SECTOR_SIZE) {
        printf("virtio: tried to read/write sector=%d, but capacity is %d\n", sector, blk_capacity / SECTOR_SIZE);
        return;
    }

    // Build a request following the specification of virtio-blk
    blk_req->sector = sector;
    blk_req->type = is_write ? VIRTIO_BLK_T_OUT : VIRTIO_BLK_T_IN;
    if (is_write) {
        memcpy(blk_req->data, buf, SECTOR_SIZE);
    }

    // Build 3 descriptors for virtual queue
    struct virtio_virtq *vq = blk_request_vq;

    vq->descs[0].addr = blk_req_paddr;
    // Read 3 fields of vertio_blk_req: sizeof(type) + sizeof(reserved) + sizeof(sector)
    vq->descs[0].len = sizeof(uint32_t) * 2 + sizeof(uint64_t);
    vq->descs[0].flags = VIRTQ_DESC_F_NEXT; // 2.7.6 Next Flag: Descriptor Chaining
    vq->descs[0].next = 1;

    // 2.6.13.1 Placing Buffers Into The Descriptor Table
    vq->descs[1].addr = blk_req_paddr + offsetof(struct virtio_blk_req, data);
    vq->descs[1].len = SECTOR_SIZE;
    vq->descs[1].flags = VIRTQ_DESC_F_NEXT | (is_write ? 0 : VIRTQ_DESC_F_WRITE);
    vq->descs[1].next = 2;

    // Write status field of virtio_blk_req
    vq->descs[2].addr = blk_req_paddr + offsetof(struct virtio_blk_req, status);
    vq->descs[2].len = sizeof(uint8_t);
    vq->descs[2].flags = VIRTQ_DESC_F_WRITE; // 2.7.3 Write Flag

    // Note: We always use the same descriptors for every request (vq->desc[0..2]) because we wait
    // for the request result with blocking.

    // Notify the virtio-blk device that there are some requests to process.
    virtq_kick(vq, 0);

    // Wait until virtio-blk device finishes to process the request.
    while (virtq_is_busy(vq)) {
    }

    // Check the request status. Non-zero means an error
    if (blk_req->status != 0) {
        printf("virtio: warn: failed to read/write sector=%d status=%d\n", sector, blk_req->status);
        return;
    }

    // Copy the data read from the device to the buffer.
    if (!is_write) {
        memcpy(buf, blk_req->data, SECTOR_SIZE);
    }
}

static int parse_octal(char const *oct, int const len) {
    int dec = 0;
    for (int i = 0; i < len; i++) {
        char const c = oct[i];
        if (c < '0' || '7' < c) {
            break;
        }

        dec = dec * 8 + (c - '0');
    }
    return dec;
}

void fs_init(void) {
    // Note: We don't use local variable for `disk` because it consumes stack. (The stack size is small)

    for (size_t sector = 0; sector < sizeof(disk) / SECTOR_SIZE; sector++) {
        read_write_disk(&disk[sector * SECTOR_SIZE], sector, false); // Read the file to `disk` variable
    }

    size_t offset = 0;
    for (int i = 0; i < FILES_MAX; i++) {
        struct tar_header const *header = (struct tar_header *)&disk[offset];
        if (header->name[0] == '\0') {
            break;
        }

        if (strcmp(header->magic, "ustar") != 0) {
            PANIC("invalid tar header: magic=\"%s\"", header->magic);
        }

        int const size = parse_octal(header->size, sizeof(header->size));
        // Add only normal files (type='0')
        if (header->type == '0') {
            struct file *file = &files[i];
            file->in_use = true;
            strcpy(file->name, header->name);
            memcpy(file->data, header->data, size);
            file->size = size;
            printf("file: %s, size=%d offset=%d\n", file->name, file->size, offset);
        }

        offset += align_up(sizeof(struct tar_header) + size, SECTOR_SIZE);
    }
}

static void unparse_octal(char *buf, int value) {
    int i = 0;
    do {
        buf[i++] = (value % 8) + '0';
        value /= 8;
    } while (value > 0);
}

// Write all files to the disk
void fs_flush(void) {
    memset(disk, 0, sizeof(disk));
    size_t offset = 0;
    for (int j = 0; j < FILES_MAX; j++) {
        struct file *file = &files[j];
        if (!file->in_use) {
            continue;
        }

        // Create tar header

        struct tar_header *header = (struct tar_header *)&disk[offset];
        memset(header, 0, sizeof(*header));
        strcpy(header->magic, "ustar");
        strcpy(header->mode, "000644");
        strcpy(header->version, "00");
        strcpy(header->name, file->name);
        header->type = '0';

        unparse_octal(header->size, file->size);

        int checksum = ' ' * sizeof(header->checksum);
        for (size_t i = 0; i < sizeof(struct tar_header); i++) {
            checksum += (unsigned char)disk[offset + i];
        }
        for (int i = 5; i >= 0; i--) {
            header->checksum[i] = (checksum % 8) + '0';
            checksum /= 8;
        }

        memcpy(header->data, file->data, file->size);
        offset += align_up(sizeof(struct tar_header) + file->size, SECTOR_SIZE);
    }

    // Write the entire on-memory data to the disk
    for (size_t sector = 0; sector < sizeof(disk) / SECTOR_SIZE; sector++) {
        read_write_disk(&disk[sector * SECTOR_SIZE], sector, true);
    }

    printf("wrote %d bytes to disk\n", sizeof(disk));
}

struct file *fs_lookup(char const *filename) {
    for (int i = 0; i < FILES_MAX; i++) {
        struct file *file = &files[i];
        if (strcmp(file->name, filename) == 0) {
            return file;
        }
    }
    return NULL;
}
