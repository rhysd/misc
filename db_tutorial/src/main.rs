use std::array;
use std::env;
use std::fmt;
use std::fs::File;
use std::io::{self, BufRead, Read, Seek, SeekFrom, Write};
use std::iter;
use std::mem;
use std::path::Path;
use std::slice;
use std::str;

enum Error {
    Unknown(String),
    Syntax(&'static str),
    TableFull,
    StringTooLong(u32),
    PageOutOfBounds(u32),
    Io(io::Error),
}

impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Unknown(cmd) => write!(f, "Unrecognized command: {cmd:?}"),
            Self::Syntax(usage) => write!(f, "Syntax error: {usage:?}"),
            Self::TableFull => write!(f, "Table is full"),
            Self::StringTooLong(max) => write!(f, "String length exceeds max length {max}"),
            Self::PageOutOfBounds(idx) => write!(f, "Page index {idx} is out of bounds"),
            Self::Io(inner) => write!(f, "I/O error: {inner}"),
        }
    }
}

type Result<T> = std::result::Result<T, Box<Error>>;

impl Error {
    #[inline]
    fn err<T>(self) -> Result<T> {
        Err(Box::new(self))
    }
}

impl From<io::Error> for Box<Error> {
    fn from(err: io::Error) -> Self {
        Box::new(Error::Io(err))
    }
}

// TODO: This must be allocated on heap since table row structure will be dynamic
#[repr(transparent)]
struct InlineString<const N: usize>([u8; N]);

impl<const N: usize> Default for InlineString<N> {
    fn default() -> Self {
        Self([0; N])
    }
}

impl<const N: usize> InlineString<N> {
    fn new(s: &str) -> Result<Self> {
        let s = s.as_bytes();
        if s.len() > N {
            return Error::StringTooLong(N as u32).err();
        }

        Ok(Self(std::array::from_fn(|i| s.get(i).copied().unwrap_or(0))))
    }

    fn as_str(&self) -> &'_ str {
        let end = self.0.iter().take_while(|&&b| b != 0).count();
        str::from_utf8(&self.0[..end]).unwrap()
    }
}

enum ReplInput<'input> {
    Meta(&'input str),
    Statement(&'input str),
}

#[derive(Default)]
struct Prompt {
    buffer: String,
}

impl Prompt {
    fn input<R: BufRead, W: Write>(
        &mut self,
        mut stdin: R,
        mut stdout: W,
    ) -> io::Result<ReplInput<'_>> {
        stdout.write_all(b"db > ")?;
        stdout.flush()?;
        self.buffer.clear();
        stdin.read_line(&mut self.buffer)?;
        let line = self.buffer.trim();
        if let Some(meta) = line.strip_prefix('.') {
            Ok(ReplInput::Meta(meta))
        } else {
            Ok(ReplInput::Statement(line))
        }
    }
}

#[derive(Debug)]
enum MetaCommand {
    Exit,
}

impl MetaCommand {
    fn parse(input: &str) -> Option<Self> {
        if input == "exit" {
            Some(Self::Exit)
        } else {
            None
        }
    }
}

#[derive(Debug)]
struct Row<'a> {
    id: u32,
    user_name: &'a str,
    email: &'a str,
}

impl<'a> fmt::Display for Row<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "({}, {:?}, {:?})", self.id, self.user_name, self.email)
    }
}

#[repr(C)]
#[repr(packed(1))]
#[derive(Default)]
struct SerializedRow {
    id: u32,                     // integer
    user_name: InlineString<32>, // varchar(32)
    email: InlineString<255>,    // varchar(255)
}

impl SerializedRow {
    const SIZE: usize = mem::size_of::<SerializedRow>();

    fn serialize(row: &Row<'_>) -> Result<Self> {
        Ok(Self {
            id: row.id,
            user_name: InlineString::new(row.user_name)?,
            email: InlineString::new(row.email)?,
        })
    }

    fn deserialize(&self) -> Row<'_> {
        Row { id: self.id, user_name: self.user_name.as_str(), email: self.email.as_str() }
    }
}
const _: () = assert!(SerializedRow::SIZE == 291);

#[derive(Default)]
struct Page(Vec<SerializedRow>);

impl Page {
    const PAGE_SIZE: usize = 4096;
    const ROWS_PER_PAGE: usize = Self::PAGE_SIZE / SerializedRow::SIZE;

    fn read_from<R: Read>(mut reader: R) -> io::Result<Self> {
        let mut rows: Vec<_> =
            iter::repeat_with(SerializedRow::default).take(Self::ROWS_PER_PAGE).collect();
        let rows_array: &mut [SerializedRow; Self::ROWS_PER_PAGE] =
            rows.as_mut_slice().try_into().unwrap();
        // Safety: This operation is safe because any bit pattern is valid for `SerializedRow` instance.
        let bytes = unsafe {
            mem::transmute::<
                &mut [SerializedRow; Self::ROWS_PER_PAGE],
                &mut [u8; mem::size_of::<[SerializedRow; Self::ROWS_PER_PAGE]>()],
            >(rows_array)
        };
        // `read_exact` is not available since the last page may not be full.
        let _ = reader.read(bytes)?;
        Ok(Self(rows))
    }

    fn write_to<W: Write>(&self, mut writer: W, num_rows: usize) -> io::Result<()> {
        assert!(num_rows <= self.0.len(), "{} is out of bounds {}", num_rows, self.0.len());
        let src_ptr = self.0.as_ptr();
        // Safety: This operation is safe because the length of array pointed by `src_ptr` and the
        // length of `bytes` are the same.
        let bytes = unsafe {
            let bytes_ptr = mem::transmute::<_, *const u8>(src_ptr);
            slice::from_raw_parts(bytes_ptr, num_rows * SerializedRow::SIZE)
        };
        writer.write_all(bytes)
    }

    fn row_at(&mut self, idx: usize) -> &mut SerializedRow {
        // `self.0` must be filled with zero-initialized SerializedRow instances. Otherwise, this
        // page read from a file will contain uninitialized instances. Accessing such instances
        // causes UB.
        if self.0.is_empty() {
            self.0.reserve_exact(Self::ROWS_PER_PAGE);
            self.0.extend(iter::repeat_with(SerializedRow::default).take(Self::ROWS_PER_PAGE));
        }
        &mut self.0[idx]
    }

    fn is_cached(&self) -> bool {
        !self.0.is_empty()
    }
}

struct Pager {
    file: Option<File>,
    file_len: u64,
    pages: [Page; Self::MAX_PAGES],
}

impl Default for Pager {
    fn default() -> Self {
        let pages = array::from_fn(|_| Page::default());
        Self { file: None, file_len: 0, pages }
    }
}

impl Pager {
    const MAX_PAGES: usize = 100;

    fn open(path: &Path) -> io::Result<Self> {
        let file =
            File::options().read(true).write(true).create(true).truncate(false).open(path)?;
        let file_len = file.metadata()?.len();
        let pages = std::array::from_fn(|_| Page::default());
        Ok(Self { file: Some(file), file_len, pages })
    }

    fn page_at(&mut self, page_num: usize) -> Result<&mut Page> {
        let Some(page) = self.pages.get_mut(page_num) else {
            return Error::PageOutOfBounds(page_num as u32).err();
        };
        if !page.is_cached() {
            if let Some(file) = &mut self.file {
                let num_pages = self.file_len as usize / Page::PAGE_SIZE
                    + usize::from(self.file_len as usize % Page::PAGE_SIZE > 0);
                if page_num <= num_pages {
                    let offset = page_num * Page::PAGE_SIZE;
                    file.seek(SeekFrom::Start(offset as u64))?;
                    *page = Page::read_from(file)?;
                }
            }
        }
        Ok(page)
    }

    fn flush(&mut self, page_num: usize, rows: usize) -> io::Result<()> {
        let Some(file) = &mut self.file else {
            return Ok(());
        };

        let Some(page) = self.pages.get_mut(page_num) else {
            return Ok(());
        };

        if !page.is_cached() {
            return Ok(());
        }

        let offset = page_num * Page::PAGE_SIZE;
        file.seek(SeekFrom::Start(offset as u64))?;
        page.write_to(file, rows)?;

        *page = Page::default();
        Ok(())
    }
}

#[derive(Default)]
struct Table {
    num_rows: u32,
    pager: Pager,
}

impl Table {
    const MAX_ROWS: usize = Page::ROWS_PER_PAGE * Pager::MAX_PAGES;

    fn open(path: &Path) -> io::Result<Self> {
        let pager = Pager::open(path)?;
        let num_rows = (pager.file_len / SerializedRow::SIZE as u64) as u32;
        Ok(Self { num_rows, pager })
    }

    fn row_slot(&mut self, row_num: u32) -> Result<&mut SerializedRow> {
        let page_num = row_num as usize / Page::ROWS_PER_PAGE;
        let page = self.pager.page_at(page_num)?;
        let row = page.row_at(row_num as usize % Page::ROWS_PER_PAGE);
        Ok(row)
    }

    fn close(&mut self) -> io::Result<()> {
        let num_full_pages = self.num_rows as usize / Page::ROWS_PER_PAGE;
        for idx in 0..num_full_pages {
            self.pager.flush(idx, Page::ROWS_PER_PAGE)?;
        }

        let num_additional_rows = self.num_rows as usize % Page::ROWS_PER_PAGE;
        if num_additional_rows > 0 {
            self.pager.flush(num_full_pages, num_additional_rows)?;
        }

        Ok(())
    }
}

#[derive(Debug)]
enum Statement<'input> {
    Insert(Row<'input>),
    Select,
}

impl<'input> Statement<'input> {
    fn prepare(input: &'input str) -> Result<Self> {
        let mut tokens = input.split_whitespace();
        let Some(cmd) = tokens.next() else {
            return Error::Unknown(String::new()).err();
        };
        match cmd {
            "insert" => {
                let mut parse = || {
                    let id = tokens.next()?.parse().ok()?;
                    let user_name = tokens.next()?;
                    let email = tokens.next()?;
                    let row = Row { id, user_name, email };
                    Some(Statement::Insert(row))
                };
                parse().ok_or_else(|| Box::new(Error::Syntax("insert {id} {user} {email}")))
            }
            "select" => Ok(Self::Select),
            c => Error::Unknown(c.to_string()).err(),
        }
    }

    fn execute<W: Write>(&self, table: &mut Table, mut w: W) -> Result<()> {
        match self {
            Self::Insert(row) => {
                if table.num_rows as usize >= Table::MAX_ROWS {
                    return Error::TableFull.err();
                }

                let slot = table.row_slot(table.num_rows)?;
                *slot = SerializedRow::serialize(row)?;
                table.num_rows += 1;

                Ok(())
            }
            Self::Select => {
                for i in 0..table.num_rows {
                    let row = table.row_slot(i)?.deserialize();
                    writeln!(w, "{row}").unwrap();
                }
                Ok(())
            }
        }
    }
}

fn repl<R: BufRead, W: Write>(mut stdin: R, mut stdout: W, mut table: Table) -> io::Result<()> {
    let mut prompt = Prompt::default();

    loop {
        match prompt.input(&mut stdin, &mut stdout)? {
            ReplInput::Meta(input) => {
                let Some(cmd) = MetaCommand::parse(input) else {
                    writeln!(stdout, "Unrecognized meta command: {input:?}").unwrap();
                    continue;
                };
                match cmd {
                    MetaCommand::Exit => {
                        table.close()?;
                        writeln!(stdout, "Bye.").unwrap();
                        break;
                    }
                }
            }
            ReplInput::Statement(input) => match Statement::prepare(input) {
                Ok(statement) => match statement.execute(&mut table, &mut stdout) {
                    Ok(()) => writeln!(stdout, "Executed: {input:?}").unwrap(),
                    Err(err) => {
                        writeln!(stdout, "Error while executing {statement:?}: {err}").unwrap();
                    }
                },
                Err(err) => writeln!(stdout, "Error while executing {input:?}: {err}").unwrap(),
            },
        }
    }

    Ok(())
}

fn main() -> io::Result<()> {
    dbg!(Page::ROWS_PER_PAGE);
    let table = if let Some(path) = env::args_os().nth(1) {
        Table::open(Path::new(&path))?
    } else {
        Table::default()
    };
    repl(io::stdin().lock(), io::stdout(), table)
}

#[cfg(test)]
#[rustfmt::skip::macros(format)]
mod tests {
    use super::*;
    use insta::assert_snapshot;
    use std::fmt::Write as _;
    use std::io::BufReader;
    use std::path::PathBuf;
    use std::sync::OnceLock;
    use tempfile::TempDir;

    #[track_caller]
    fn run_test_with_table(stdin: impl AsRef<str>, table: Table) -> io::Result<String> {
        let stdin = stdin.as_ref();
        let mut stdout = Vec::<u8>::new();
        let mut stdin = BufReader::new(stdin.as_bytes());
        repl(&mut stdin, &mut stdout, table)?;
        Ok(String::from_utf8(stdout).unwrap())
    }

    #[track_caller]
    fn run_test(stdin: impl AsRef<str>) -> io::Result<String> {
        run_test_with_table(stdin.as_ref(), Table::default())
    }

    #[test]
    fn single_row_insert_select() {
        let input = "\
            insert 1 user1 person@example.com
            select
            .exit
        ";
        assert_snapshot!(run_test(input).unwrap());
    }

    #[test]
    fn rows_more_than_one_page() {
        let mut s = String::new();
        for i in 1..=15 {
            writeln!(s, "insert {i} user{i} person{i}@example.com").unwrap();
        }
        s.push_str("select\n");
        s.push_str(".exit\n");
        assert_snapshot!(run_test(s).unwrap());
    }

    #[test]
    fn table_full_error() {
        let mut s = String::new();
        for i in 1..=1401 {
            writeln!(s, "insert {i} user{i} person{i}@example.com").unwrap();
        }
        s.push_str(".exit\n");
        assert_snapshot!(run_test(s).unwrap());
    }

    #[test]
    fn max_user_name_and_email() {
        let user = "a".repeat(32);
        let email = "a".repeat(255);
        let input = format!("\
            insert 1 {user} {email}
            select
            .exit
        ");
        assert_snapshot!(run_test(input).unwrap());
    }

    #[test]
    fn name_is_too_long() {
        let user = "a".repeat(33);
        let input = format!("\
            insert 1 {user} foo@example.com
            select
            .exit
        ");
        assert_snapshot!(run_test(input).unwrap());
    }

    #[test]
    fn email_is_too_long() {
        let email = "a".repeat(256);
        let input = format!("\
            insert 1 foo {email}
            select
            .exit
        ");
        assert_snapshot!(run_test(input).unwrap());
    }

    #[test]
    fn id_must_not_be_negative() {
        let input = "\
            insert -1 foo foo@example.com
            select
            .exit
        ";
        assert_snapshot!(run_test(input).unwrap());
    }

    fn temp_dir() -> &'static TempDir {
        static TEMP_DIR: OnceLock<TempDir> = OnceLock::new();
        TEMP_DIR.get_or_init(|| tempfile::tempdir().unwrap())
    }

    fn temp_file(name: &str) -> PathBuf {
        temp_dir().path().join(name)
    }

    #[track_caller]
    fn run_test_persistent(name: &str, stdin: impl AsRef<str>) -> io::Result<String> {
        let path = temp_file(name);
        let table = Table::open(&path)?;
        run_test_with_table(stdin.as_ref(), table)
    }

    #[test]
    fn persistent_single_row() {
        let mut output = run_test_persistent(
            "single_row.db",
            "insert 1 foo foo@example.com
            .exit",
        )
        .unwrap();
        output += &run_test_persistent(
            "single_row.db",
            "select
            .exit",
        )
        .unwrap();
        assert_snapshot!(output);
    }

    #[test]
    fn persistent_single_page() {
        let mut input = String::new();
        for i in 1..=14 {
            writeln!(input, "insert {i} user{i} user{i}@example.com").unwrap();
        }
        input.push_str(".exit\n");
        let mut output = run_test_persistent("single_page.db", input).unwrap();
        output += &run_test_persistent(
            "single_page.db",
            "select
            .exit",
        )
        .unwrap();
        assert_snapshot!(output);
    }

    #[test]
    fn persistent_single_and_half_page() {
        let mut input = String::new();
        for i in 1..=21 {
            writeln!(input, "insert {i} user{i} user{i}@example.com").unwrap();
        }
        input.push_str(".exit\n");
        let mut output = run_test_persistent("single_half_page.db", input).unwrap();
        output += &run_test_persistent(
            "single_half_page.db",
            "select
            .exit",
        )
        .unwrap();
        assert_snapshot!(output);
    }

    #[test]
    fn persistent_no_row() {
        let mut output = run_test_persistent("nothing.db", ".exit").unwrap();
        output += &run_test_persistent(
            "nothing.db",
            "select
            .exit",
        )
        .unwrap();
        assert_snapshot!(output);
    }
}
