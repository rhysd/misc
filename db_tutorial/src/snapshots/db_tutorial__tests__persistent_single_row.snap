---
source: src/main.rs
expression: output
---
db > Executed: "insert 1 foo foo@example.com"
db > Bye.
db > (1, "foo", "foo@example.com")
Executed: "select"
db > Bye.
